File: tags/2.0-M4/trunk/mobileCloud/rimos/mobileObject/src/main/java/org/openmobster/core/mobileCloud/rimos/module/mobileObject/MobileObjectDatabase.java
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.openmobster.core.mobileCloud.rimos.module.mobileObject;

import java.util.Vector;
import java.util.Enumeration;

import org.openmobster.core.mobileCloud.rimos.errors.SystemException;
import org.openmobster.core.mobileCloud.rimos.module.bus.Bus;
import org.openmobster.core.mobileCloud.rimos.module.bus.Invocation;
import org.openmobster.core.mobileCloud.rimos.module.bus.InvocationResponse;
import org.openmobster.core.mobileCloud.rimos.service.Registry;
import org.openmobster.core.mobileCloud.rimos.service.Service;
import org.openmobster.core.mobileCloud.rimos.storage.DBException;
import org.openmobster.core.mobileCloud.rimos.storage.Database;
import org.openmobster.core.mobileCloud.rimos.storage.Record;
import org.openmobster.core.mobileCloud.rimos.util.GenericAttributeManager;


/**
 * @author openmobster@gmail.com
 */
public final class MobileObjectDatabase extends Service
{		
	public static String storageId = "storageId";
	public static String recordId = "recordId";
	public static String syncOperation = "syncOperation";
	
	public MobileObjectDatabase()
	{	
	}
	
	public void start()
	{
		
	}
	
	public void stop()
	{
		
	}
	
	public static MobileObjectDatabase getInstance()
	{
		return (MobileObjectDatabase)Registry.getInstance().lookup(MobileObjectDatabase.class);
	}
	//---------------------------------------------------------------------------------------------------------------------------------------------	
	public MobileObject read(String storageId, String recordId)
	{
		try
		{
			MobileObject mobileObject = null;
						
			Vector all = this.readAll(storageId);
			Enumeration cour = all.elements();
			while(cour.hasMoreElements())
			{
				MobileObject curr = (MobileObject)cour.nextElement();
				
				if(curr.getRecordId().equals(recordId))
				{
					return curr;
				}
			}
		
			return mobileObject;
		}
		catch(Exception e)
		{
			throw new SystemException(this.getClass().getName(), "read", new Object[]{
				storageId, recordId,
				"Exception="+e.toString(),
				"Error="+e.getMessage()
			});
		}
	}
	
	public Vector readAll(String storageId)
	{	
		try
		{
			Vector objects = new Vector();
						
			this.checkStorage(storageId);
			Enumeration all = Database.getInstance().selectAll(storageId);
			while(all.hasMoreElements())
			{
				Record record = (Record)all.nextElement();
				objects.addElement(new MobileObject(record));
			}
			
			return objects;
		}
		catch(Exception e)
		{
			throw new SystemException(this.getClass().getName(), "readAll", new Object[]{
				storageId,
				"Exception="+e.toString(),
				"Error="+e.getMessage()
			});
		}
	}
	
	public String create(MobileObject mobileObject)
	{
		try
		{
			String newId = null;
			
			//Map the objectId locally to the device
			Invocation invocation = new Invocation(mobileObject.getStorageId());
			invocation.setValue("storageId", mobileObject.getStorageId());
			invocation.setValue("recordId", mobileObject.getRecordId());
			InvocationResponse response = Bus.getInstance().invokeService(invocation);			
			if(response != null)
			{
				String mappedRecordId = response.getValue(InvocationResponse.returnValue);
				if(mappedRecordId != null && mappedRecordId.trim().length()>0)
				{
					mobileObject.setRecordId(mappedRecordId);
				}
			}						
			
			this.checkStorage(mobileObject.getStorageId());
			newId = Database.getInstance().insert(mobileObject.getStorageId(), mobileObject.getRecord());									
			
			return newId;
		}
		catch(Exception be)
		{			
			throw new SystemException(this.getClass().getName(), "create", new Object[]
   			{
				"storageId="+mobileObject.getStorageId(),
				"recordId="+mobileObject.getRecordId(),
   				"error="+be.getMessage()
   			}
   			);
		}		
	}
		
	public void update(MobileObject mobileObject)
	{
		try
		{
			this.checkStorage(mobileObject.getStorageId());
			
			Record recordToBeUpdated = mobileObject.getRecord();
			
			String dirtyStatus = recordToBeUpdated.getDirtyStatus();
			if(dirtyStatus == null || dirtyStatus.trim().length() == 0)
			{
				Record currentRecord = Database.getInstance().select(mobileObject.getStorageId(), mobileObject.getRecordId());
				recordToBeUpdated.setDirtyStatus(currentRecord.getDirtyStatus());
			}
			
			Database.getInstance().update(mobileObject.getStorageId(), recordToBeUpdated);
		}
		catch(Exception e)
		{
			throw new SystemException(this.getClass().getName(), "update", new Object[]
   			{
				"storageId="+mobileObject.getStorageId(),
				"recordId="+mobileObject.getRecordId(),
   				"error="+e.getMessage()
   			}
   			);
		}		
	}
	
	public void delete(MobileObject mobileObject)
	{	
		try
		{
			this.checkStorage(mobileObject.getStorageId());
			
			MobileObject objectToBeDeleted = this.read(mobileObject.getStorageId(), mobileObject.getRecordId());
			Record recordToBeDeleted = objectToBeDeleted.getRecord();
			
			Database.getInstance().delete(mobileObject.getStorageId(), recordToBeDeleted);
		}
		catch(Exception e)
		{
			throw new SystemException(this.getClass().getName(), "delete", new Object[]
 			{
				"storageId="+mobileObject.getStorageId(),
				"recordId="+mobileObject.getRecordId(),
				"error="+e.getMessage()
 			}
 			);
		}		
	}
	
	public void deleteAll(String storageId)
	{
		try
		{
			this.checkStorage(storageId);
			Database.getInstance().deleteAll(storageId);
		}
		catch(Exception e)
		{
			throw new SystemException(this.getClass().getName(), "deleteAll", new Object[]
   			{
				"storageId="+storageId,
				"error="+e.getMessage()
   			}
   			);
		}		
	}
	//---Query Integration------------------------------------------------------------------------------------------------------------------------------------------
	public Vector query(String storageId, GenericAttributeManager queryAttributes)
	{
		Vector result = this.readAll(storageId);
		
		int logicLink = LogicChain.AND; //assumed by default
		if(queryAttributes.getAttribute("logicLink")!=null)
		{
			logicLink = ((Integer)queryAttributes.getAttribute("logicLink")).intValue();
		}
		
		Vector expressions = (Vector)queryAttributes.getAttribute("expressions");
		if(expressions != null && !expressions.isEmpty())
		{
			LogicChain chain = null;
			switch(logicLink)
			{
				case LogicChain.AND:
					chain = LogicChain.createANDChain();					
				break;
				
				case LogicChain.OR:
					chain = LogicChain.createORChain();
				break;
				
				default:
				break;
			}
			if(chain != null)
			{
				int size = expressions.size();
				for(int i=0; i<size; i++)
				{
					chain.add((LogicExpression)expressions.elementAt(i));
				}
				Query query = Query.createInstance(chain);
				result = query.executeQuery(result);
			}
		}
		
		return result;
	}
	//----------------------------------------------------------------------------------------------------------------------------------------------
	public void syncWithServer(String storageId)
	{
		
	}	
	//----------------------------------------------------------------------------------------------------------------------------------------------
	private void checkStorage(String storageId) throws DBException
	{
		if(!Database.getInstance().doesTableExist(storageId))
		{
			Database.getInstance().createTable(storageId);
		}
	}
	//-----------------------------------------------------------------------------------------------------------------------------------
	
}

File: trunk/iphone/mobilecloudlib/MobileObjectDatabase.m
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

#import "MobileObjectDatabase.h"
#import "LogicChain.h"
#import "LogicExpression.h"
#import "Query.h"
#import "MetaData.h"


/**
 * 
 * @author openmobster@gmail.com
 */
@implementation MobileObjectDatabase

+(MobileObjectDatabase *) getInstance
{
	Registry *registry = [Registry getInstance];
	return (MobileObjectDatabase *)[registry lookup:[MobileObjectDatabase class]];
}

-(NSArray *)readAll:(NSString *)channel
{
	NSArray *persistentMobileObjects = [PersistentMobileObject findByChannel:channel];
	NSMutableArray *all = [NSMutableArray array];
	if(all != nil)
	{
		for(PersistentMobileObject *local in persistentMobileObjects)
		{
			[all addObject:[local parseMobileObject]];
		}
	}
	return [NSArray arrayWithArray:all];
}

-(MobileObject *)read:(NSString *)channel:(NSString *)recordId
{
    PersistentMobileObject *stored = [PersistentMobileObject findByOID:recordId];
    if(stored != nil)
    {
        MobileObject *mobileObject = [stored parseMobileObject];
        return mobileObject;
    }
	return nil;
}

-(NSString *)create:(MobileObject *)mobileObject
{
	PersistentMobileObject *newObject = [PersistentMobileObject newInstance:mobileObject.service];
	
	mobileObject.dirtyStatus = [GeneralTools generateUniqueId];
	
	[newObject setState:mobileObject];
    
	
    BOOL success = [newObject saveInstance];
    if(!success)
    {
        NSMutableArray *params = [NSMutableArray arrayWithObjects:@"commit_failure",nil];
        SystemException *ex = [SystemException withContext:@"MobileObjectDatabase" method:@"create" parameters:params];
        @throw ex; 
    }
	
    return newObject.oid;
}

-(void)update:(MobileObject *)mobileObject
{
	PersistentMobileObject *stored = [PersistentMobileObject findByOID:mobileObject.recordId];
	if(stored != nil)
	{
		//Check for outdated data (Optimistic Locking)
		NSString *storedStatus = stored.dirtyStatus;
		NSString *activeStatus = mobileObject.dirtyStatus;
		
		if(activeStatus != nil)
		{
			if(![storedStatus isEqualToString:activeStatus])
			{
				NSMutableArray *params = [NSMutableArray arrayWithObjects:@"Optimistic Locking Failure!!",nil];
				SystemException *ex = [SystemException withContext:@"MobileObjectDatabase" method:@"update" parameters:params];
				@throw ex;
			}
		}
		
		mobileObject.dirtyStatus = [GeneralTools generateUniqueId];
		[stored setState:mobileObject];
        
        
		BOOL success = [stored saveInstance];
        if(!success)
        {
            NSMutableArray *params = [NSMutableArray arrayWithObjects:@"commit_failure",nil];
            SystemException *ex = [SystemException withContext:@"MobileObjectDatabase" method:@"update" parameters:params];
            @throw ex; 
        }
	}
}

-(void)delete:(MobileObject *)mobileObject
{
	PersistentMobileObject *stored = [PersistentMobileObject findByOID:mobileObject.recordId];
	if(stored != nil)
	{
		BOOL success = [PersistentMobileObject delete:stored];
        if(!success)
        {
            NSMutableArray *params = [NSMutableArray arrayWithObjects:@"commit_failure",nil];
            SystemException *ex = [SystemException withContext:@"MobileObjectDatabase" method:@"delete" parameters:params];
            @throw ex; 
        }
	}
}

-(void)deleteAll:(NSString *)channel
{
	BOOL success = [PersistentMobileObject deleteAll:channel];
    if(!success)
    {
        NSMutableArray *params = [NSMutableArray arrayWithObjects:@"commit_failure",nil];
        SystemException *ex = [SystemException withContext:@"MobileObjectDatabase" method:@"deleteAll" parameters:params];
        @throw ex; 
    }
}

-(NSSet *)query:(NSString *)channel:(GenericAttributeManager *)queryAttributes
{
    //Find the Logic Link
    int logicLink = AND; //default value
    NSNumber *logicLinkAttribute = [queryAttributes getAttribute:@"logicLink"];
    if(logicLinkAttribute != nil)
    {
        logicLink = [logicLinkAttribute intValue];
    }
    
    //Setup the expressions
    NSArray *expressions = (NSArray *)[queryAttributes getAttribute:@"expressions"];
    if(expressions == nil || [expressions count] == 0)
    {
        NSMutableArray *params = [NSMutableArray arrayWithObjects:@"Query structure is improper!!",nil];
        SystemException *sys = [SystemException withContext:@"MobileObjectDatabase" method:@"query" parameters:params];
        @throw sys;
    }
    
    //Establish the LogicChain
    LogicChain *chain = nil;
    switch(logicLink)
    {
        case AND:
            chain = [LogicChain createANDChain];					
            break;
            
        case OR:
            chain = [LogicChain createORChain];
            break;
            
        default:
            break;
    }
    if(chain == nil)
    {
        NSMutableArray *params = [NSMutableArray arrayWithObjects:@"Query structure is improper!!",nil];
        SystemException *sys = [SystemException withContext:@"MobileObjectDatabase" method:@"query" parameters:params];
        @throw sys;
    }
    
    //Now get possible matches of records for each expression
    NSMutableSet *result = [NSMutableSet set];
    for(LogicExpression *courExpr in expressions)
    {
        [chain addExpression:courExpr];
        
        //get beans that match this expression
        NSSet *matchedBeans = [self logicExpressionBeans:channel :courExpr];
        if(matchedBeans != nil)
        {
            [result unionSet:matchedBeans];
        }
    }
    
    Query *query = [Query createInstance:chain];
    NSSet *queryResults = [query executeQuery:result];
    
    return queryResults;
}

-(NSSet *)logicExpressionBeans:(NSString *) channel :(LogicExpression *) expression
{
    NSMutableSet *result = [NSMutableSet set];
    
    NSString *lhs = expression.lhs;
    NSString *rhs = expression.rhs;
    
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"PersistentMobileObject" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    NSString *lhsExpression = [NSString stringWithFormat:@"name=%@",lhs];
    NSString *rhsExpression = [NSString stringWithFormat:@"value=%@",rhs];
    
    NSArray *stored = nil;
    if(expression.op == OP_EQUALS || expression.op == OP_NOT_EQUALS)
    {
        if(expression.op == OP_EQUALS)
        {
            //EqualToQuery
            NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(service == %@) AND (nameValuePairs contains %@) AND (nameValuePairs contains %@)",channel, lhsExpression, rhsExpression];
            [request setPredicate:predicate];
        }
        else if(expression.op == OP_NOT_EQUALS)
        {
            //NotEqualToQuery
            NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(service == %@) AND (nameValuePairs contains %@) AND (NONE nameValuePairs contains %@)",channel, lhsExpression, rhsExpression];  
            [request setPredicate:predicate];
        }
        
        stored = [managedContext executeFetchRequest:request error:NULL];
        
        //filter by predicate
        if(stored != nil)
        {
            for(PersistentMobileObject *local in stored)
            {
                [result addObject:[local parseMobileObject]];
            }
        }
    }
    else
    {
        //In case of LIKE and CONTAINS queries
        NSArray *all = [self readAll:channel];
        for(MobileObject *local in all)
        {
            [result addObject:local];
        }
    }

    return result;
}

-(BOOL) isBooted:(NSString *)channel
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"PersistentMobileObject" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"oid" ascending:YES];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    //Set the predicate
    //find by channel
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(service == %@)",channel];
    [request setPredicate:predicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
    managedObjectContext:managedContext sectionNameKeyPath:nil
    cacheName:nil];
    cursor = [cursor autorelease];
    
    [cursor performFetch:NULL];
    //NSLog(success ? @"Success: Yes" : @"Success: No");
    
    NSArray *results = cursor.fetchedObjects;
    if(results != nil && [results count]>0)
    {
        return YES;
    }
    
    return NO;
}

-(NSFetchedResultsController *)searchExactMatchAND:(NSString *)channel :(GenericAttributeManager *)criteria
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"PersistentMobileObject" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"service" ascending:NO];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    //Do some predicate magic here
    NSMutableArray *channelPredicate = [NSMutableArray array];
    NSMutableArray *criteriaPredicate = [NSMutableArray array];
    
    //Channel Predicate
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(service == %@)",channel];
    [channelPredicate addObject:predicate];
    
    if(criteria != nil)
    {
        NSArray *names = [criteria getNames];
        for(NSString *name in names)
        {
            NSString *value = [criteria getAttribute:name];
            NSPredicate *nameValuePredicate = [NSPredicate predicateWithFormat:@"(nameValuePairs contains %@) AND (nameValuePairs contains %@)",name,value];
            [criteriaPredicate addObject:nameValuePredicate];
        }
    }
    
    NSPredicate *channelCompound = [NSCompoundPredicate andPredicateWithSubpredicates:[NSArray arrayWithArray:channelPredicate]];
    NSPredicate *criteriaCompound = [NSCompoundPredicate andPredicateWithSubpredicates:[NSArray arrayWithArray:criteriaPredicate]];
    
    NSMutableArray *all = [NSMutableArray array];
    [all addObject:channelCompound];
    [all addObject:criteriaCompound];
    NSArray *allPredicates = [NSArray arrayWithArray:all];
    NSPredicate *finalPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:allPredicates];
    
    [request setPredicate:finalPredicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                        managedObjectContext:managedContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    cursor = [cursor autorelease];
    
    //Now perform fetch
    NSError *error;
    BOOL success = [cursor performFetch:&error];
    if(success)
    {
        return cursor;
    }
    else
    {
        NSLog(@"SearchExactMatchAND Error: %@",[error userInfo]);
    }

    return nil;
}

-(NSFetchedResultsController *)searchExactMatchOR:(NSString *)channel :(GenericAttributeManager *)criteria
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"PersistentMobileObject" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"service" ascending:NO];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    //Do some predicate magic here
    NSMutableArray *channelPredicate = [NSMutableArray array];
    NSMutableArray *criteriaPredicate = [NSMutableArray array];
    
    //Channel Predicate
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(service == %@)",channel];
    [channelPredicate addObject:predicate];
    
    if(criteria != nil)
    {
        NSArray *names = [criteria getNames];
        for(NSString *name in names)
        {
            NSString *value = [criteria getAttribute:name];
            NSPredicate *nameValuePredicate = [NSPredicate predicateWithFormat:@"(nameValuePairs contains %@) AND (nameValuePairs contains %@)",name,value];
            [criteriaPredicate addObject:nameValuePredicate];
        }
    }
    
    NSPredicate *channelCompound = [NSCompoundPredicate andPredicateWithSubpredicates:[NSArray arrayWithArray:channelPredicate]];
    NSPredicate *criteriaCompound = [NSCompoundPredicate orPredicateWithSubpredicates:[NSArray arrayWithArray:criteriaPredicate]];
    
    NSMutableArray *all = [NSMutableArray array];
    [all addObject:channelCompound];
    [all addObject:criteriaCompound];
    NSArray *allPredicates = [NSArray arrayWithArray:all];
    NSPredicate *finalPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:allPredicates];
    
    [request setPredicate:finalPredicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                        managedObjectContext:managedContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    cursor = [cursor autorelease];
    
    //Now perform fetch
    NSError *error;
    BOOL success = [cursor performFetch:&error];
    if(success)
    {
        return cursor;
    }
    else
    {
        NSLog(@"SearchExactMatchOR Error: %@",[error userInfo]);
    }
    
    return nil;  
}

-(NSFetchedResultsController *)readByName:(NSString *)channel :(NSString *)name :(BOOL) ascending
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"MetaData" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"value" ascending:ascending];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    NSPredicate *channelPredicate = [NSPredicate predicateWithFormat:@"(parent.service == %@)",channel];
    NSPredicate *namePredicate = [NSPredicate predicateWithFormat:@"(name == %@)",name];
    NSArray *predicates = [NSArray arrayWithObjects:channelPredicate,namePredicate,nil];
    
    NSPredicate *finalPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
    
    [request setPredicate:finalPredicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                        managedObjectContext:managedContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    cursor = [cursor autorelease];
    
    //Now perform fetch
    NSError *error;
    BOOL success = [cursor performFetch:&error];
    if(success)
    {
        return cursor;
    }
    else
    {
        NSLog(@"readByNameAndSort Error: %@",[error userInfo]);
    }

    
    return nil;
}

-(NSFetchedResultsController *)readByName:(NSString *)channel :(NSString *)name
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"MetaData" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"name" ascending:YES];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    NSPredicate *channelPredicate = [NSPredicate predicateWithFormat:@"(parent.service == %@)",channel];
    NSPredicate *namePredicate = [NSPredicate predicateWithFormat:@"(name == %@)",name];
    NSArray *predicates = [NSArray arrayWithObjects:channelPredicate,namePredicate,nil];
    
    NSPredicate *finalPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
    
    [request setPredicate:finalPredicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                        managedObjectContext:managedContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    cursor = [cursor autorelease];
    
    //Now perform fetch
    NSError *error;
    BOOL success = [cursor performFetch:&error];
    if(success)
    {
        return cursor;
    }
    else
    {
        NSLog(@"readByName Error: %@",[error userInfo]);
    }
    
    
    return nil;
}

-(NSFetchedResultsController *)readProxyCursor:(NSString *)channel
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"PersistentMobileObject" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"service" ascending:NO];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    //Channel Predicate
    NSPredicate *channelPredicate = [NSPredicate predicateWithFormat:@"(service == %@)",channel];
    NSPredicate *proxyPredicate = [NSPredicate predicateWithFormat:@"(proxy == %@)",[NSNumber numberWithBool:YES]];
    
    NSArray *predicates = [NSArray arrayWithObjects:channelPredicate,proxyPredicate,nil];
    NSPredicate *finalPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
    [request setPredicate:finalPredicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                        managedObjectContext:managedContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    cursor = [cursor autorelease];
    
    //Now perform fetch
    NSError *error;
    BOOL success = [cursor performFetch:&error];
    if(success)
    {
        return cursor;
    }
    else
    {
        NSLog(@"readProxyCursor Error: %@",[error userInfo]);
    }
    
    return nil;
}

-(NSFetchedResultsController *)readByNameValuePair:(NSString *)channel :(NSString *)name :(NSString *)value
{
    //Get the Storage Context
    NSManagedObjectContext *managedContext = [[CloudDBManager getInstance] storageContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"MetaData" 
                                              inManagedObjectContext:managedContext];
    
    //Get an instance if its already been provisioned
    NSFetchRequest *request = [[[NSFetchRequest alloc] init] autorelease];
    [request setEntity:entity];
    
    //Set the cursor size
    [request setFetchBatchSize:1];
    
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"name" ascending:YES];
    NSArray *descriptors = [[NSArray alloc] initWithObjects:descriptor, nil];
    [request setSortDescriptors:descriptors];
    [descriptor release];
    [descriptors release];
    
    NSPredicate *channelPredicate = [NSPredicate predicateWithFormat:@"(parent.service == %@)",channel];
    NSPredicate *namePredicate = [NSPredicate predicateWithFormat:@"(name == %@)",name];
    NSPredicate *valuePredicate = [NSPredicate predicateWithFormat:@"(value == %@)",value];
    NSArray *predicates = [NSArray arrayWithObjects:channelPredicate,namePredicate,valuePredicate,nil];
    
    NSPredicate *finalPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
    
    [request setPredicate:finalPredicate];
    
    NSFetchedResultsController *cursor =
    [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                        managedObjectContext:managedContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    cursor = [cursor autorelease];
    
    //Now perform fetch
    NSError *error;
    BOOL success = [cursor performFetch:&error];
    if(success)
    {
        return cursor;
    }
    else
    {
        NSLog(@"readByNameValuePair Error: %@",[error userInfo]);
    }
    
    
    return nil; 
}
@end
